### PR TITLE
#4297 new vaccination history column

### LINKED ISSUE #4297: New column for vaccination history
## Is your feature request related to a problem? Please describe.

Display the name notes created here: https://github.com/openmsupply/mobile/issues/4296 in the vaccination step 1 page.

## Describe the solution you'd like

Just display the name notes of patient event `🤷 ` in the table?

## Implementation

Unsure- I think there might be an issue with receiving all the name notes from the patient api

## Describe alternatives you've considered

- Showing this detail in the current `patient history` modal. Users want to see this in step1, and also might be too cluttered?

## Additional context
- https://github.com/sussol/msupply/issues/9253
- Might be an issue for historic users - what about tonga? Will they just have the column and not show any details? 🤔

### PR DESCRIPTION
**Pull request by joshxg**
Fixes #4297 

## Change summary

- Adds a history column and modal
image

- Modal (oops, mind the data- the last column with actual data is what will show)
image

- Went with this for now- a lot easier..
image

## Testing

- [ ] Can view a patients vaccine history by pressing the history icon

### Related areas to think about

- Went with the history icon- could be better.
- Didn't disable rows when there isn't any history was easier not to- would have to pass through all the name notes of all the patients in the list and check if there is a name note with the patient event of 'vaccination'
- Not really sure the table format is needed, could be a lot nicer
- `PatientSelect` is now massive and should spend a bit of time to clean it up

### CODE REVIEW

File: src/actions/Entities/VaccinePrescriptionActions.js [hunk 1]
```diff
@@ -179,10 +179,11 @@ const createVaccinationNameNote = (
   const data = {
     refused,
     bonusDose,
-    vaccine: selectedBatch?.itemName,
-    vaccineCode: selectedBatch?.itemCode,
+    itemName: selectedBatch?.itemName,
+    itemCode: selectedBatch?.itemCode,
```
Comment (joshxg) [inline src/actions/Entities/VaccinePrescriptionActions.js:183]: easier to use these names.. 😁

File: src/actions/Entities/VaccinePrescriptionActions.js [hunk 2]
```diff
@@ -179,10 +179,11 @@ const createVaccinationNameNote = (
...
     batch: selectedBatch?.batch,
     expiry: selectedBatch?.expiryDate,
+    vaccineDate: moment().format('DD/MM/YYYY'), // Duplicating entry date for ease of reporting
```
Comment (joshxg) [inline src/actions/Entities/VaccinePrescriptionActions.js:186]: im lazy?

File: src/pages/dataTableUtilities/getColumns.js [hunk 3]
```diff
@@ -135,6 +137,14 @@ const PAGE_COLUMNS = {
     COLUMN_NAMES.PRESCRIPTION_DATE,
     COLUMN_NAMES.PRESCRIBER,
   ],
+
+  [MODALS.VACCINATION_HISTORY]: [
+    COLUMN_NAMES.ITEM_CODE,
+    COLUMN_NAMES.ITEM_NAME,
+    COLUMN_NAMES.VACCINATOR,
+    COLUMN_NAMES.VACCINE_DATE,
+  ],
```
Comment (joshxg) [inline src/pages/dataTableUtilities/getColumns.js:146]: Could remove all the `vaccine history` stuff now?

File: src/pages/dataTableUtilities/getColumns.js [hunk 4]
```diff
@@ -163,6 +173,19 @@ const PAGE_COLUMNS = {
...
+    COLUMN_NAMES.LAST_NAME,
+    COLUMN_NAMES.DATE_OF_BIRTH,
+    COLUMN_NAMES.PATIENT_HISTORY,
+  ],
+  [MODALS.VACCINE_PATIENT_LOOKUP]: [
+    COLUMN_NAMES.FIRST_NAME,
+    COLUMN_NAMES.LAST_NAME,
+    COLUMN_NAMES.DATE_OF_BIRTH,
+    COLUMN_NAMES.PATIENT_HISTORY,
+  ],
```
Comment (joshxg) [inline src/pages/dataTableUtilities/getColumns.js:187]: should probably add the history to the normal lookup also 🤷

File: src/widgets/DataTable/DataTable.js [hunk 5]
```diff
@@ -106,17 +112,19 @@ const DataTable = React.memo(
...
+            ref={virtualizedListRef}
+            keyboardDismissMode="none"
+            data={data}
+            keyboardShouldPersistTaps="always"
+            style={style}
+            ListFooterComponent={<KeyboardSpacing />}
+            renderItem={renderItem}
+            {...otherProps}
+          />
+        </View>
```
Comment (joshxg) [inline src/widgets/DataTable/DataTable.js:127]: this was for putting the table in `Paper` which wasn''t playing nice. Seems to be OK everywhere

File: src/widgets/DataTable/DataTableHeaderRow.js [hunk 6]
```diff
@@ -41,7 +41,7 @@ const DataTableHeaderRow = React.memo(({ columns, sortKey, isAscending, onPress
             columnKey={key}
             onPress={sortable ? onPress : null}
             sortDirection={directionForThisColumn}
-            sortable={sortable}
+            sortable={isSortable ? sortable : false}
```
Comment (joshxg) [inline src/widgets/DataTable/DataTableHeaderRow.js:44]: the columns of the patient select weren't originally sortable, and I didn't want to bother spending time to do it

File: src/widgets/SimpleTable.js [hunk 7]
```diff
@@ -123,6 +122,7 @@ export const SimpleTable = React.memo(
           ListHeaderComponent={renderHeaders}
           extraData={selectedRows}
           style={style}
+          {...flatListProps}
```
Comment (joshxg) [inline src/widgets/SimpleTable.js:125]: means I can override `keyExtractor` as the vaccine history rows don't have an `id` - all the uses of `SimpleTable` seem to be OK

File: src/widgets/modals/VaccinationHistory.js [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+
+const jsonSchema = {
+  type: 'object',
+  properties: {
+    bonusDose: {
+      type: 'boolean',
+    },
+    refused: {
+      type: 'boolean',
+    },
```
Comment (joshxg) [inline src/widgets/modals/VaccinationHistory.js:28]: should maybe display `bonusDose` or `refused`?

**Review by mark-prins — APPROVED** [review]
Looks really good! 👍 
One change required - The name note should be using the `vaccination` code now

File: src/widgets/Tabs/PatientSelect.js [hunk 9]
```diff
@@ -211,6 +212,60 @@ const PatientSelectComponent = ({
...
+    () => <DataTableHeaderRow columns={columns} isSortable={false} />,
+    [columns]
+  );
+
+  const getCallback = colKey => {
+    switch (colKey) {
+      case 'patientHistory':
+        return patientId => {
+          const [vaccinationPatientEvent] = UIDatabase.objects('PatientEvent').filtered(
+            "code == 'RV'"
```
[the remark this answers is not in the document]
Comment (joshxg) [reply src/widgets/Tabs/PatientSelect.js]: 8892308

**Comment by joshxg** [comment]
Also added  53cfa6f and 44359e1 to not show the column when the patient event isnt there

**Pull request merged**